stdlib/random.ml

```diff
@@ -114,35 +114,106 @@ module State = struct
   let make_self_init () =
     make (random_seed ())
 
-  (* Return 30 random bits as an integer 0 <= x < 1073741824 *)
+  (* NOTE:
+   *   0x3FFF_FFFF = 2^30 - 1, which is max_int for 31-bit integers
+   *   -0x4000_0000 = -2^30, which is min_int for 31-bit integers
+   *   0x7FFF_FFFF = 2^31 - 1, which is the "all-ones" 31-bit integer
+   *)
+
+  (* Return 30 random bits as an integer 0 <= x < 2^30 *)
   let bits s =
     Int64.to_int (next s) land 0x3FFF_FFFF
 
-  (* Return an integer between 0 (included) and [bound] (excluded) *)
-  let rec intaux s n =
+  (* Return an integer between 0 (included) and [bound] (excluded).
+     The bound must fit in 31 bits.
+     This function is designed so that it yields the same output
+     on 32-bit and 64-bit platforms. *)
+  let rec int31aux s n =
     let r = bits s in
     let v = r mod n in
-    if r - v > 0x3FFFFFFF - n + 1 then intaux s n else v
+    (* Explanation of this test: see comment in [full_int_aux]. *)
+    if r - v > 0x3FFFFFFF - n + 1 then int31aux s n else v
 
   let int s bound =
     if bound > 0x3FFFFFFF || bound <= 0
     then invalid_arg "Random.int"
-    else intaux s bound
+    else int31aux s bound
 
   (* Return an integer between 0 (included) and [bound] (excluded).
      [bound] may be any positive [int]. *)
-  let rec int63aux s n =
+  let rec full_int_aux s n =
+    (* We start by drawing any representable non-negative integer. *)
     let r = Int64.to_int (next s) land max_int in
     let v = r mod n in
-    if r - v > max_int - n + 1 then int63aux s n else v
+    (* For uniform distribution of the result between 0 included and [n]
+     * excluded, the random number [r] must have been drawn uniformly in
+     * an interval whose length is a multiple of [n]. To achieve this,
+     * we use rejection sampling on the greatest interval [ [0, k*n-1] ]
+     * that fits in representable integers, that is, we reject the
+     * sample if it falls outside of this interval, and draw again.
+     * This is what the test below does (while carefuly avoiding
+     * overflows and sparing a division [max_int / n]). *)
+    if r - v > max_int - n + 1 then full_int_aux s n else v
 
   let full_int s bound =
     if bound <= 0 then
       invalid_arg "Random.full_int"
+    (* When the bound fits in 31 bits, we use [int31aux]
+     * so as to yield the same output on 32-bit and 64-bit platforms. *)
     else if bound > 0x3FFFFFFF then
-      int63aux s bound
+      full_int_aux s bound
+    else
+      int31aux s bound
+
+  (* Return an integer between [min] (included) and [max] (included).
+     We must have [min <= max], and [min] and [max] must fit in 31 bits.
+     This function is designed so that it yields the same output on
+     32-bit and 64-bit platforms. *)
+  let rec int31_in_range_aux s ~min ~max =
+    let r = Int64.to_int (next s) land 0x7FFF_FFFF in
+    if r < min || r > max then int31_in_range_aux s ~min ~max else r
```

Comment:
Regarding 32-bit platforms: I’m confused. I don’t have a 32-bit system at hand, but I verified on my 64-bit system that OCaml (toplevel/compiler) accepts the equivalent constant `0x7FFF_FFFF_FFFF_FFFF` and maps it to `-1`. Is the behavior different for 32-bit OCaml?

Regarding 64-bit platforms: indeed, my mistake.

New plan to reduce the sample to 31 bits: keep the 30 lowest bits, and replicate the sign bit in the remaining higher bits.

```ocaml
    (* 0x3FFF_FFF = 2^30 - 1 has its 30 lowest bits equal to one.
     * -0x4000_000 = -2^30 has all bits equal to one except the 30 lowest bits which are equal to zero.
     *)
    let r = Int64.to_int (next s) in
    let r = (if r >= 0 then r land 0x3FFF_FFFF else (r land 0x3FFF_FFFF) lor -0x4000_0000) in
```

or, without branching:

```ocaml
    let r = Int64.to_int (next s) in
    let s = r asr Sys.int_size in (* this replicates the sign bit of [r] to all bits *)
    let r = (s land -0x4000_0000) lor (r land 0x3FFF_FFFF) in
```

Does that seem correct to you?

Reply:
1-  OCaml on a 32-bit platform will indeed accept 0x7FFF_FFFF and read it as -1.  However, if you compile to bytecode on a 64-bit platform, 0x7FFF_FFFF reads as 2^31-1, and that's the constant that would be put in the bytecode executable, making it impossible to run it on a 32-bit platform.  That's what the "Generated bytecode executable "ocamlc" cannot be used on a 32-bit platform" error means.

2- Sign extension of the low 31 bits sound like a good idea, but can be achieved much more easily:
```
let d = Sys.int_size - 31 in (r lsl d) asr d
```

Reply:
1. Thanks for the explanation!
2. That’s much better indeed. By contrast with my version, yours also has the good taste of computing, in fact, the signed remainder modulo 2^31 of its input, even when `int`s have more than 31 bits (because yours replicate the 31th bit whereas mine replicate the highest bit); so it makes sense to isolate this code in a function `int31_of_int`. Regarding the constant `d`, do I need to write it in some way so that the compiler optimizes it to an immediate constant operand for the shift?

Reply:
I applied the suggested fix, and I also added (numerous) chi2 tests.

[For future readers (since the commit has now disappeared): the faulty code was:
```ocaml
  let rec int31_in_range_aux s ~min ~max =
    let r = Int64.to_int (next s) land 0x7FFF_FFFF in (* WRONG *)
    if r < min || r > max then int31_in_range_aux s ~min ~max else r
```
]